pkg/ring/kv/memberlist/metrics.go

```diff
@@ -0,0 +1,206 @@
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "messages_in_broadcast_queue",
+		Help:      "Number of user messages in the broadcast queue",
+	}, func() float64 {
+		return float64(m.broadcasts.NumQueued())
+	})
+
+	m.totalSizeOfBroadcastMessagesInQueue = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "messages_in_broadcast_size_total",
+		Help:      "Total size of messages waiting in the broadcast queue",
+	})
+
+	m.casAttempts = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "cas_attempt_count",
+		Help:      "Attempted CAS operations",
+	})
+
+	m.casSuccesses = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "cas_success_count",
+		Help:      "Successful CAS operations",
+	})
+
+	m.casFailures = prometheus.NewCounter(prometheus.CounterOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "cas_failure_count",
+		Help:      "Failed CAS operations",
+	})
+
+	m.storeValuesDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(m.cfg.MetricsNamespace, subsystem, "kv_store_count"),
+		"Number of values in KV Store",
+		nil, nil,
+	)
+
+	m.storeSizesDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(m.cfg.MetricsNamespace, subsystem, "kv_store_value_size"),
+		"Sizes of values in KV Store in bytes",
+		[]string{"key"}, nil)
+
+	m.memberlistMembersCount = prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "cluster_members_count",
+		Help:      "Number of members in memberlist cluster",
+	}, func() float64 {
+		return float64(m.memberlist.NumMembers())
+	})
+
+	m.memberlistHealthScore = prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "cluster_node_health_score",
+		Help:      "Health score of this cluster. Lower the better. 0 = totally health",
+	}, func() float64 {
+		return float64(m.memberlist.GetHealthScore())
+	})
+
+	m.memberlistMembersInfoDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(m.cfg.MetricsNamespace, subsystem, "cluster_members_last_timestamp"),
+		"State information about memberlist cluster members",
+		[]string{"name", "address"}, nil)
+
+	m.watchPrefixDroppedNotifications = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: m.cfg.MetricsNamespace,
+		Subsystem: subsystem,
+		Name:      "watch_prefix_dropped_notifications",
+		Help:      "Number of dropped notifications in WatchPrefix function",
+	}, []string{"prefix"})
+
+	if m.cfg.MetricsRegisterer == nil {
+		return
+	}
+
+	// if registration fails, that's too bad, but don't panic
+	for _, c := range m.getAllMetrics() {
+		err := m.cfg.MetricsRegisterer.Register(c.(prometheus.Collector))
```

Comment:
Why not panic? It shouldn't be possible to pass the same register to different MemberListClients and if it does happen it should happen relatively quickly, correct?

Reply:
You're right. This came from failing unit tests that tried to register the same metrics, but then I added MetricsRegisterer to the config, which fixed the problem as well. This isn't necessary anymore, same as unregister.